### PR TITLE
Fix polling period, handle extensions to run in extension slice and log extension run cgroup details

### PR DESCRIPTION
**Pull request by nagworld9**
## Description

> Check resource metrics polling is within configured period
> Handle extensions to starts in extension slice if they are already installed.
> Log extension run cgroup details when it's starts

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/osutil/default.py [hunk 1]
```diff
@@ -1521,3 +1521,23 @@ def _run_command_raising_OSUtilError(cmd, err_msg, cmd_input=None):
                 "{0}, Retcode: {1}, Output: {2}, Error: {3}".format(err_msg, e.returncode, e.stdout, e.stderr))
         except Exception as e:
             raise OSUtilError("{0}, Retcode: {1}, Error: {2}".format(err_msg, -1, ustr(e)))
+
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/osutil/default.py:1524]: or maybe it is... did you try those command-line options on different distros?

File: azurelinuxagent/common/osutil/default.py [hunk 2]
```diff
@@ -1521,3 +1521,23 @@ def _run_command_raising_OSUtilError(cmd, err_msg, cmd_input=None):
...
+        try:
+            command = ['ps', '-opid', '--no-headers', '--ppid', str(pid)]
+            output = shellutil.run_command(command)
+            return output
+        except Exception as error:
+            logger.warn("Error retrieving child processes: {0}".format(ustr(error)))
+            return ""
+
+    def get_extension_process_id(self, pid):
+        # Extension run is child of the systemd-run and grandchild of the shell process.
```
Comment (narrieta) [inline azurelinuxagent/common/osutil/default.py:1537]: why the mention to systemd-run?

File: azurelinuxagent/common/osutil/default.py [hunk 3]
```diff
@@ -1521,3 +1521,23 @@ def _run_command_raising_OSUtilError(cmd, err_msg, cmd_input=None):
...
+        # This method returns child processes of the parent.
+        try:
+            command = ['ps', '-opid', '--no-headers', '--ppid', str(pid)]
+            output = shellutil.run_command(command)
+            return output
+        except Exception as error:
+            logger.warn("Error retrieving child processes: {0}".format(ustr(error)))
+            return ""
+
+    def get_extension_process_id(self, pid):
```
Comment (narrieta) [inline azurelinuxagent/common/osutil/default.py:1536]: the cgroup configurator may be a better place for this function

File: azurelinuxagent/common/osutil/default.py [hunk 4]
```diff
@@ -1521,3 +1521,23 @@ def _run_command_raising_OSUtilError(cmd, err_msg, cmd_input=None):
...
+        # Extension run is child of the systemd-run and grandchild of the shell process.
+        ps_output = self._get_child_processes(pid)
+        child_process_ids = [int(line) for line in ps_output.splitlines()]
+        if child_process_ids:
+            ps_output = self._get_child_processes(child_process_ids[0])
```
Comment (narrieta) [inline azurelinuxagent/common/osutil/default.py:1541]: why are we using item 0 only?

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 5]
```diff
@@ -771,6 +772,9 @@ def setup_extension_slice(self, extension_name):
                 try:
                     slice_contents = _EXTENSION_SLICE_CONTENTS.format(extension_name=extension_name)
                     CGroupConfigurator._Impl.__create_unit_file(extension_slice_path, slice_contents)
+
+                    logger.info("Executing systemctl daemon-reload...")
+                    shellutil.run_command(["systemctl", "daemon-reload"])
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupconfigurator.py:777]: did you verify this is actually needed?

File: azurelinuxagent/common/cgroupapi.py [hunk 6]
```diff
@@ -275,6 +275,12 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
 
             # We start systemd-run with shell == True so process.pid is the shell's pid, not the pid for systemd-run
             self._systemd_run_commands.append(process.pid)
+            # Extension run is child of the systemd-run and grandchild of the shell process.
+            ext_run_pid = osutil.get_extension_process_id(process.pid)
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupapi.py:279]: we have a race condition here: the extension may not be started yet when Popen returns

File: azurelinuxagent/ga/exthandlers.py [hunk 7]
```diff
@@ -1411,9 +1411,8 @@ def _enable_extension(self, extension, uninstall_exit_code):
         env = {
             ExtCommandEnvVariable.UninstallReturnCode: uninstall_exit_code
         }
-        # This check to call the setup if AzureMonitorLinuxAgent extension already installed and not called setup before
-        if self.is_azuremonitorlinuxagent(self.get_full_name()) and \
```
Comment (narrieta) [inline azurelinuxagent/ga/exthandlers.py:1415]: why are we removing the check for AMA?

File: azurelinuxagent/ga/monitor.py [hunk 8]
```diff
@@ -51,14 +51,19 @@ class PollResourceUsage(PeriodicOperation):
...
         super(PollResourceUsage, self).__init__(conf.get_cgroup_check_period())
         self.__log_metrics = conf.get_cgroup_log_metrics()
+        self._current_run_time = datetime.datetime.utcnow()
 
     def _operation(self):
         tracked_metrics = CGroupsTelemetry.poll_all_tracked()
 
-        for metric in tracked_metrics:
-            report_metric(metric.category, metric.counter, metric.instance, metric.value, log_event=self.__log_metrics)
+        previous_run_time = self._current_run_time
```
Comment (narrieta) [inline azurelinuxagent/ga/monitor.py:59]: not sure we want to skip this; let's sync offline

File: azurelinuxagent/common/osutil/default.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nagworld9) [reply azurelinuxagent/common/osutil/default.py:1524]: I tested few different distros and options are available.

Seems like have other command without need for extra command-line options to eliminate headers. I'm thinking to change to this.

```
suse12:/home/azureuser # pgrep --parent 1490
1576
1578
```

**Pull request closed without merging**